FEAT: elastic search integration

Junjiequan opened the pull request:
## Description

ElasticSearch integration for SciCat as optional feature. 
Disable elasticSearch will fall back to the regular mongoDB query.

## Motivation

MongoDB's aggregation method can yield low search accuracy. Longer queries often result in less relevant outcomes, frustrating users.

In contrast, Elasticsearch boasts a robust search engine. It indexes and stores data flexibly, allowing adaptable search methods based on user preferences. It promptly presents the most relevant information using document relevancy scores, significantly enhancing user experiences.




## Tests included/Docs Updated?

- [x] Included for each change/fix?
- [ ] Passing? (Merge will not be approved unless this is checked) 
- [ ] Docs updated?
- [x] New packages used/requires npm install? 
- [ ] Toggle added for new features?

bpedersen2 committed:
Add message for instrument creation problems

Change-Id: I1d91d7ea3a66cdec62d9c667eeebc30d5bd1e30b

Junjiequan committed:
feat: elastic search MVP integration

Junjiequan committed:
feat: fulltext query implementation

Junjiequan committed:
Fixed elastic search dto and added hard filter

Junjiequan committed:
[WIP] Feat: Elastic search integration

Junjiequan committed:
Fix: some clean ups and added api endpoint

Junjiequan committed:
feat: added error handler for ES services

Junjiequan committed:
fix: update local docker compose with elastic search

Junjiequan committed:
fix: fixed prepare:local script

Junjiequan committed:
fix: include elastic search env file for local environment

Junjiequan committed:
trigger code check

Junjiequan committed:
Merge remote-tracking branch 'origin/master' into SWAP-3409-elastic-search-install-and-configure-elastic-sear

Junjiequan committed:
fix: minor cleanup

Junjiequan committed:
fix: fix unit tests

enthusiastio commented:
Let me know if I should look into it while it is WIP

Junjiequan commented:
> Let me know if I should look into it while it is WIP

@enthusiastio This PR is ready for review now, I will change description by tomorrow : )

Junjiequan committed:
fix: minor cleanup

Junjiequan committed:
fix: include missing variable into the ENV

Junjiequan committed:
fix: update test setup

Junjiequan committed:
fix: specify env file path for docker compose

Junjiequan committed:
fix: fixed  E2E test

Junjiequan committed:
test

Junjiequan committed:
test

Junjiequan committed:
revert to previous tess yml file

Junjiequan committed:
fix: fix dependency issue

Junjiequan committed:
fix: enable normal search when ES search is disabled

Junjiequan committed:
fix: prevent application crashing from failed ES connection

Junjiequan committed:
fix: minor fix for query filter & some cleanup

Junjiequan committed:
fix: minor code cleanup

enthusiastio commented on package.json:
```diff
@@ -27,14 +27,16 @@
     "test:api:jest": "jest --config ./test/config/jest-e2e.json --maxWorkers=50%",
     "test:api:mocha": "mocha --config ./test/config/.mocharc.json",
-    "prepare:local": "docker-compose -f CI/E2E/docker-compose-local.yaml up -d && cp functionalAccounts.json.example functionalAccounts.json"
+    "prepare:local": "docker-compose -f CI/E2E/docker-compose-local.yaml --env-file CI/E2E/.env.elastic-search up  -d && cp functionalAccounts.json.example functionalAccounts.json"
```
I wonder if we didn t want to store the environmental variables in HELM charts, if there is any difference between environments

Junjiequan replied:
@enthusiastio
We should store the ENV variables in the helm charts for dev environment. This prepare:local script is only for local usage, to spin up containers for local development.

enthusiastio reviewed (approved):
To be honest not everything I comprehend, as I don; t have that much experience with scat, but to me overall this looks good, just a small comment

nitrosx commented:
@Junjiequan I'm going to review it by tomorrow.

bolmsten committed:
Merge branch 'master' into SWAP-3409-elastic-search-install-and-configure-elastic-sear

Junjiequan committed:
Merge branch 'master' into SWAP-3409-elastic-search-install-and-configure-elastic-sear

Junjiequan committed:
Feat: sync elastic search data with mongoDB CRUD operation  (#717)

* feat: update ES data along with mongoDB CRUD actions

* feat: sync ES data with mongoDB on CRUD operation

* fix: minor text change

Junjiequan committed:
Merge branch 'master' into SWAP-3409-elastic-search-install-and-configure-elastic-sear

Junjiequan committed:
fix: minor cleanup

nitrosx commented on CI/E2E/.env.elastic-search:
```diff
@@ -0,0 +1,18 @@
+ES_USERNAME=elastic
+ES_PASSWORD=password
+KIBANA_PORT=5601
+KIBANA_PASSWORD=password
```
Do we need all of them?

nitrosx commented on CI/E2E/backend.env:
```diff
@@ -34,3 +34,21 @@ CREATE_DATASET_WITH_PID_GROUPS=""
+ES_USERNAME=elastic
+ES_PASSWORD=password
+KIBANA_PORT=5601
+KIBANA_PASSWORD=password
```
Do we need all of them here?

Junjiequan replied:
I'll cleanup these, only keeping necessary ones

nitrosx commented on src/common/utils.ts:
```diff
@@ -480,14 +481,17 @@ export const createFullfacetPipeline = <T, Y extends object>(
 ): PipelineStage[] => {
   const pipeline = [];
-  const facetMatch: Record<string, unknown> = {};
+  const facetMatch: Record<string, unknown> = esPids
```
I am wondering if we are going to see some performance issues if the list of pids in esPIDS becomes really long

Junjiequan replied:
Fetching 20K records seems no performance issue yet.
But I have included a condition for where we need to fetch all data. if there's no text in the search query, elastic search will not be executed and we use mongo query to retrieve all data.
```
   if (!this.ESClient || !$text) {
      const datasets = await this.datasetModel
        .find(whereClause, null, modifiers)
        .exec();
      return datasets;
    }
```

nitrosx commented on src/config/configuration.ts:
```diff
@@ -158,6 +158,29 @@ const configuration = () => {
+      enabled: process.env.KIBANA_ENABLED ?? "no",
+      port: parseInt(process.env.KIBANA_PORT || "5601", 10),
+      password: process.env.KIBANA_PASSWORD,
+    },
```
Do we need all this entries for elastic search?
Why do we need kibana configuration in scicat backend?

Junjiequan replied:
I will cleanup this

nitrosx commented on src/datasets/datasets.service.ts:
```diff
@@ -41,22 +43,33 @@ import { DatasetClass, DatasetDocument } from "./schemas/dataset.schema";
       addCreatedByFields(createDatasetDto, username),
     );
-
+    if (this.ESClient) {
+      this.ESClient.insertDocument(
+        createdDataset.toObject() as DatasetDocument,
+      );
+    }
```
How do we change which dataset fields we are inserting in elastic search?

Junjiequan replied:
```
  async insertDocument(data: DatasetDocument) {
    delete data._id;
    try {
      await this.esService.index({
        index: this.defaultIndex,
        id: data.pid,
        document: data,
      });
```
I think we need to specify here, let the `document: ` to only update given field.
Should be feasible to insert or update only specific field

nitrosx replied:
WE should specify the document fields that will be inserted in ES in scicat BE configuration.
Let's open a new issue regarding this topic

nitrosx commented on src/datasets/datasets.service.ts:
```diff
@@ -88,16 +100,33 @@ export class DatasetsService {
     };
     const modifiers: QueryOptions = parseLimitFilters(filter.limits);
+    const { $text, ...remainingClauses } = whereClause;
+    if (!this.ESClient || !$text) {
+      const datasets = await this.datasetModel
+        .find(whereClause, null, modifiers)
+        .exec();
+      return datasets;
+    }
+
+    const esResult = await this.ESClient.search(
+      filter.fields as IDatasetFields,
+      modifiers.limit,
+      modifiers.skip,
+    );
 
     const datasets = await this.datasetModel
-      .find(whereClause, null, modifiers)
+      .find(
+        { _id: { $in: esResult.data }, ...remainingClauses },
+        null,
+        modifiers,
+      )
```
It seems to me that here the elastic search is always used in the search.
What happens if we disable elastic search in configuration?

Junjiequan replied:
```
    if (!this.ESClient || !$text) {
      const datasets = await this.datasetModel
        .find(whereClause, null, modifiers)
        .exec();
      return datasets;
    }
```
if we disable elastic search `!this.ESClient` will be true and it will execute only within the block

nitrosx commented on src/datasets/datasets.service.ts:
```diff
@@ -355,6 +419,15 @@ export class DatasetsService {
+      throw new NotFoundException();
+    }
+  }
+
```
What does this function do?

Junjiequan replied:
elastic search itself generates `_id` for each document, so we need to remove `_id` that we get from mongo in order to do the data sync.
But since we already have PID field, we can simply specify { _id:0 } to retrive data from mongoDB without id

nitrosx replied:
Could would you please insert a comment that this function is used only when a manual sync is triggered.

nitrosx commented on src/elastic-search/elastic-search.controller.ts:
```diff
@@ -0,0 +1,138 @@
+import {
```
We need to implement permissions?
Do we need access to the underlying elastic search from scicat?

Junjiequan replied:
Yes we login to elastic search from scicat, but I don't think we need to implement permissions. Once elastic search is enabled anyone should be able to search through elastic search. And let the Datasource controller or service layer to decide what data to show which group of people - that being said, there's still improvement needed to achieve that

nitrosx replied:
Check how dataset controller implements the authorization.
When I check the code the ES endpoint did not specify any authorization on any of them!!!

nitrosx commented on src/elastic-search/mappings/datasetFieldMapping.ts:
```diff
@@ -0,0 +1,67 @@
+    fields: {
+      keyword: {
+        type: "keyword",
+        ignore_above: 256,
```
Do we really limit the search to 256 characters?

Junjiequan replied:
Type keyword means that the text will be stored as a whole which can only be searched by the entire text. So, unless an user want to search by writing an article, character limits under 256 should be more than enough

nitrosx commented on src/elastic-search/query-builder.service.ts:
```diff
@@ -0,0 +1,98 @@
+            },
+          },
+        })),
+      );
```
Can we submit the query as it is to elastic search without splitting it in terms?

Junjiequan replied:
We can do that, but then we lose the benefit of partial match with multiple terms to list results based on relevancy score.
One solution would be we provide a toggle button on the filter UI and let the user to decide whether they want to splitting terms or not.

nitrosx replied:
Either way is fine by me, but if we split the query, we have to do something more elaborate according to NLP methodologies!!!
Have you checked if ES can perform the query splitting internally?

nitrosx committed:
Merge branch 'master' into SWAP-3409-elastic-search-install-and-configure-elastic-sear

bpedersen2 committed:
Use separate eslint/prettier step

and remove posttest in package json.

That way lint errors are much more visible in the test stages output.

Change-Id: Icd0383cecd43277805fa9eea4ce5d221dda11b72

bpedersen2 commented on CI/E2E/docker-compose-local.yaml:
I would keep a non-es local compose as well

bpedersen2 commented on .github/workflows/test.yml:
Probably needs an e2e test with elasticsearch enabled as well

nitrosx replied:
I agree. Let's open an issue to address this

bpedersen2 commented on CI/E2E/docker-compose-local.yaml:
```diff
@@ -16,12 +16,139 @@ services:
+
+  esSetup:
+    image: docker.elastic.co/elasticsearch/elasticsearch:${STACK_VERSION}
+    volumes:
```
I would setup dev/test without TLS

Junjiequan committed:
Merge branch 'master' into SWAP-3409-elastic-search-install-and-configure-elastic-sear

Junjiequan committed:
fix: retry if ES connection fails

Junjiequan committed:
test

Junjiequan committed:
test: run test with elastic search enabled

Junjiequan committed:
fix: included unit testing for ES service & opimized ES folder structure

Junjiequan committed:
fix: remove console logs

Junjiequan committed:
fix: test e2e

Junjiequan committed:
chore: cleanup env & docker compose

Junjiequan committed:
chore: update github action with es container

Junjiequan committed:
chore: run api tests with elastic search

Junjiequan committed:
chore: remove port exposing of ES on github action

Junjiequan committed:
fix: added missing ENV in github action

Junjiequan committed:
fix: add missing ENV for test.yaml

Junjiequan committed:
Merge branch 'master' into SWAP-3409-elastic-search-install-and-configure-elastic-sear

Junjiequan committed:
fix: remove ports exposure of ES in e2e environment

Junjiequan committed:
fix: create and sync dataset collection on application init

Junjiequan committed:
fix: remove data sync from application initialization

Junjiequan committed:
fix: minor cleanup

bpedersen2 committed:
Add prettier check shortcut format:check

Change-Id: Iedba153243f0613361e761067b24a28d89f0536e

bpedersen2 committed:
Fix eslintrc as recommended

Fix rules flagged by `npx eslint-config-prettier `

See https://github.com/prettier/eslint-config-prettier#what-and-why

Change-Id: I3441f316e92358f7f9c1ee0bd8eb0956f608d3a3

Junjiequan committed:
fix: should not search empty space string

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/parser from 6.6.0 to 6.7.0

Bumps [@typescript-eslint/parser](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/parser) from 6.6.0 to 6.7.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/parser/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v6.7.0/packages/parser)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/parser"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump mongoose from 7.5.0 to 7.5.2

Bumps [mongoose](https://github.com/Automattic/mongoose) from 7.5.0 to 7.5.2.
- [Release notes](https://github.com/Automattic/mongoose/releases)
- [Changelog](https://github.com/Automattic/mongoose/blob/master/CHANGELOG.md)
- [Commits](https://github.com/Automattic/mongoose/compare/7.5.0...7.5.2)

---
updated-dependencies:
- dependency-name: mongoose
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Junjiequan committed:
Merge branch 'master' into SWAP-3409-elastic-search-install-and-configure-elastic-sear

Junjiequan committed:
fix: minor refactoring

dependabot[bot] committed:
build(deps-dev): bump @types/node-fetch from 2.6.4 to 2.6.5

Bumps [@types/node-fetch](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node-fetch) from 2.6.4 to 2.6.5.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node-fetch)

---
updated-dependencies:
- dependency-name: "@types/node-fetch"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump @nestjs/passport from 10.0.1 to 10.0.2

Bumps [@nestjs/passport](https://github.com/nestjs/passport) from 10.0.1 to 10.0.2.
- [Release notes](https://github.com/nestjs/passport/releases)
- [Changelog](https://github.com/nestjs/passport/blob/master/.release-it.json)
- [Commits](https://github.com/nestjs/passport/compare/10.0.1...10.0.2)

---
updated-dependencies:
- dependency-name: "@nestjs/passport"
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump @nestjs/config from 3.0.1 to 3.1.1

Bumps [@nestjs/config](https://github.com/nestjs/config) from 3.0.1 to 3.1.1.
- [Release notes](https://github.com/nestjs/config/releases)
- [Changelog](https://github.com/nestjs/config/blob/master/.release-it.json)
- [Commits](https://github.com/nestjs/config/compare/3.0.1...3.1.1)

---
updated-dependencies:
- dependency-name: "@nestjs/config"
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump @user-office-software/duo-message-broker

Bumps [@user-office-software/duo-message-broker](https://github.com/UserOfficeProject/user-office-lib) from 1.5.2 to 1.6.0.
- [Commits](https://github.com/UserOfficeProject/user-office-lib/commits)

---
updated-dependencies:
- dependency-name: "@user-office-software/duo-message-broker"
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump uuid and @types/uuid

Bumps [uuid](https://github.com/uuidjs/uuid) and [@types/uuid](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/uuid). These dependencies needed to be updated together.

Updates `uuid` from 9.0.0 to 9.0.1
- [Changelog](https://github.com/uuidjs/uuid/blob/main/CHANGELOG.md)
- [Commits](https://github.com/uuidjs/uuid/compare/v9.0.0...v9.0.1)

Updates `@types/uuid` from 9.0.3 to 9.0.4
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/uuid)

---
updated-dependencies:
- dependency-name: uuid
  dependency-type: direct:production
  update-type: version-update:semver-patch
- dependency-name: "@types/uuid"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump docker/build-push-action from 4 to 5

Bumps [docker/build-push-action](https://github.com/docker/build-push-action) from 4 to 5.
- [Release notes](https://github.com/docker/build-push-action/releases)
- [Commits](https://github.com/docker/build-push-action/compare/v4...v5)

---
updated-dependencies:
- dependency-name: docker/build-push-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @types/nodemailer from 6.4.9 to 6.4.10

Bumps [@types/nodemailer](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/nodemailer) from 6.4.9 to 6.4.10.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/nodemailer)

---
updated-dependencies:
- dependency-name: "@types/nodemailer"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump docker/metadata-action from 4 to 5

Bumps [docker/metadata-action](https://github.com/docker/metadata-action) from 4 to 5.
- [Release notes](https://github.com/docker/metadata-action/releases)
- [Upgrade guide](https://github.com/docker/metadata-action/blob/master/UPGRADE.md)
- [Commits](https://github.com/docker/metadata-action/compare/v4...v5)

---
updated-dependencies:
- dependency-name: docker/metadata-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Junjiequan committed:
fix: added userGroups check with search query

dependabot[bot] committed:
build(deps-dev): bump sinon from 15.2.0 to 16.0.0

Bumps [sinon](https://github.com/sinonjs/sinon) from 15.2.0 to 16.0.0.
- [Release notes](https://github.com/sinonjs/sinon/releases)
- [Changelog](https://github.com/sinonjs/sinon/blob/main/docs/changelog.md)
- [Commits](https://github.com/sinonjs/sinon/compare/v15.2.0...v16.0.0)

---
updated-dependencies:
- dependency-name: sinon
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Junjiequan committed:
fix: refactor query-builder class

dependabot[bot] committed:
build(deps-dev): bump @types/node from 20.6.1 to 20.6.3

Bumps [@types/node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node) from 20.6.1 to 20.6.3.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node)

---
updated-dependencies:
- dependency-name: "@types/node"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump docker/setup-buildx-action from 2 to 3

Bumps [docker/setup-buildx-action](https://github.com/docker/setup-buildx-action) from 2 to 3.
- [Release notes](https://github.com/docker/setup-buildx-action/releases)
- [Commits](https://github.com/docker/setup-buildx-action/compare/v2...v3)

---
updated-dependencies:
- dependency-name: docker/setup-buildx-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump docker/login-action from 2 to 3

Bumps [docker/login-action](https://github.com/docker/login-action) from 2 to 3.
- [Release notes](https://github.com/docker/login-action/releases)
- [Commits](https://github.com/docker/login-action/compare/v2...v3)

---
updated-dependencies:
- dependency-name: docker/login-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Junjiequan committed:
fix: included userGuards for elastic search controllers

Junjiequan committed:
fix: es-lint test

Junjiequan committed:
fix: lint fix

Junjiequan committed:
fix: fix lint issue

Junjiequan committed:
Merge branch 'master' into SWAP-3409-elastic-search-install-and-configure-elastic-sear

Junjiequan committed:
fix: lint issue fix

Junjiequan committed:
Merge branch 'master' into SWAP-3409-elastic-search-install-and-configure-elastic-sear

bpedersen2 commented:
I would prefer if we squash (or manually rebase and squash in smaller groups) to get a cleaner history before merging.
